fvm/environment/parse_restricted_checker.go

```diff
@@ -0,0 +1,37 @@
+package environment
+
+import (
+	"github.com/onflow/cadence/runtime/common"
+
+	"github.com/onflow/flow-go/fvm/errors"
+	"github.com/onflow/flow-go/fvm/state"
+)
+
+// Utility struct used for checking unexpected operation access while cadence
+// is parsing programs.
+type parseRestrictedChecker struct {
+	txnState *state.TransactionState
+}
+
+func (checker parseRestrictedChecker) parseRestricted(op string) error {
+	if checker.txnState.IsParseRestricted() {
+		return errors.NewParseRestrictedModeInvalidAccessFailure(op)
+	}
+
+	return nil
+}
+
+func (checker parseRestrictedChecker) getAddressValue(
+	op string,
```

Comment:
nit: maybe in future we can have opCode instead of string. but no strong preference here.